File: opencv-cpp/captureui/src/CaptureUI.cpp
/*
 * Copyright (c) Steven P. Goldsmith. All rights reserved.
 *
 * Created by Steven P. Goldsmith on November 2, 2015
 * sgoldsmith@codeferm.com
 */

#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

/**
 * A simple video capture app.
 *
 * argv[1] = camera index, url or will default to "-1" if no args passed.
 *
 * @author sgoldsmith
 * @version 1.0.0
 * @since 1.0.0
 */
int main(int argc, char *argv[]) {
	int return_val = 0;
	cout << CV_VERSION << endl;
	cout << "Press [Esc] to exit" << endl;
	VideoCapture capture;
	Mat frame;
	// See if URL arg passed
	if (argc == 2) {
		char *end;
		long l = strtol(argv[1], &end, 10);
		// See if long conversion worked
		if (end != argv[1] || *end == '\0') {
			// Use number version
			cout << "URL: " << l << endl;
			capture.open(l);
		} else {
			// Use string version
			cout << "URL: " << argv[1] << endl;
			capture.open(argv[1]);
		}
	} else {
		// Use default device
		cout << "URL: -1" << endl;
		capture.open(-1);
	}
	// See if video capture opened
	if (capture.isOpened()) {
		cout << "Resolution: " << capture.get(CAP_PROP_FRAME_WIDTH) << "x"
				<< capture.get(CAP_PROP_FRAME_HEIGHT) << endl;
		bool exit_loop = false;
		// Display frames until escape pressed
		while (capture.read(frame) && !exit_loop) {
			if (!frame.empty()) {
				imshow("C++ Capture", frame);
				int c = waitKey(10);
				// Wait for escape
				if ((char) c == 27) {
					exit_loop = true;
				}
			} else {
				cout << "No frame captured" << endl;
				exit_loop = true;
			}
		}
		// Release VideoCapture
		capture.release();
	} else {
		cout << "Unable to open device" << endl;
		return_val = -1;
	}
	return return_val;
}

File: opencv-python/codeferm/captureui.py
"""
Copyright (c) Steven P. Goldsmith. All rights reserved.

Created by Steven P. Goldsmith on December 21, 2013
sgjava@gmail.com
"""

"""A simple video capture script using imshow.

sys.argv[1] = camera index, url or will default to "-1" if no args passed.

@author: sgoldsmith

"""
# Configure logger

import logging, sys, re, socket, cv2

logger = logging.getLogger("CaptureUI")
logger.setLevel("INFO")
formatter = logging.Formatter("%(asctime)s %(levelname)-8s %(module)s %(message)s")
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(formatter)
logger.addHandler(handler)
# If no args passed then use default camera
if len(sys.argv) < 2:
    url = -1
# If arg is an integer then convert to int
elif re.match(r"[-+]?\d+$", sys.argv[1]) is not None:
    url = int(sys.argv[1])
else:
    url = sys.argv[1]
# Set socket timeout in seconds (defaults to forever)
socket.setdefaulttimeout(10)    
videoCapture = cv2.VideoCapture()
# This returns True even for bad URLs
success = videoCapture.open(url)
logger.info("OpenCV %s" % cv2.__version__)
logger.info("Press [Esc] to exit")
logger.info("URL: %s" % url)
logger.info("Resolution: %dx%d" % (videoCapture.get(cv2.CAP_PROP_FRAME_WIDTH),
                               videoCapture.get(cv2.CAP_PROP_FRAME_HEIGHT)))
# Deal with VideoCapture always returning True otherwise it will hang on VideoCapture.grab()
if videoCapture.get(cv2.CAP_PROP_FRAME_WIDTH) > 0 and videoCapture.get(cv2.CAP_PROP_FRAME_HEIGHT) > 0:
    cv2.namedWindow("Python Capture")
    key = -1
    # Wait for escape to be pressed
    while(key != 27 and success):
        videoCapture.grab()
        success, image = videoCapture.read()
        cv2.imshow("Python Capture", image)
        key = cv2.waitKey(1)
    cv2.destroyAllWindows()
else:
    logger.error("Unable to open device")
